lisa/sut_orchestrator/qemu/schema.py

```diff
@@ -0,0 +1,23 @@
+from dataclasses import dataclass
+from typing import Optional
+
+from dataclasses_json import dataclass_json
+
+
+@dataclass_json()
+@dataclass
+class CloudInitSchema:
+    extra_user_data: Optional[str] = None
+
+
+@dataclass_json()
+@dataclass
+class QemuPlatformSchema:
+    pass
```

Comment:
Do you have a design about the platform schema? In my mind, it should list all available servers, and its resources (cpu, memory). So the LISA can manage the resource to maximize the test capability. We can start from simple one like below,

```python
class QemuServer(schema.NodeSpace):
    #  the resource is available for testing. For example, if a server has 24 cpus, there may be 22. The rest 2 cpus is left for host.
    # rather than what the node space supported. Extend it to qemu specified settings on server level.
    ...

class QemuPlatformSchema:
    servers: List[QemuServer]

```

Reply:
Currently, the VMs are just created on the host that is running the test suite. And the VMs are created using a fixed size.

In a subsequent change, I will add support for dynamic VM sizes based on the minimum test requirements. Though I will probably just assume the host has the resources required to run the tests that are executed and let the tests fail if it doesn't.

I don't currently have any plans to support remote/multiple VM hosts or maximizing resource usage. Though we can add these features in the future if there absence is ever felt.

Reply:
1. Resource capacity on one host. It's ok to unlimited for now. It doesn't break the schema, if we add it later. LISA supports run test cases in parallel, it can speed up test a lot. If no capability supported, we may limit it to 1 or 2 to avoid out of resource.
2. Multiple VM hosts. It may be a break change in future, if we want to support it later. I'm not clear the test cases, which plan to run in qemu. Some test cases like cross host network traffic, live migration needs multiple hosts.
3. Remote host supports. It's ok for now. If the lisa can noot run remotely, it may not be flexible to schedule runs in ADO.

Reply:
RE 2: If we add the server list in the future, if the list is empty then we can assume that the current host should be used. Otherwise, we use the list of servers provided. So, I don't think there is a breaking change concern.

RE 3: You can use custom agents in ADO that support nested VMs. We already do this for some of the Mariner+Kubernetes testing.